csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,599 @@
+        /// </summary>
+        public const long ArrowFlagDictionaryOrdered = 1;
+        /// <summary>
+        /// For dictionary-encoded types, whether the ordering of dictionary indices is semantically meaningful.
+        /// </summary>
+        public const long ArrowFlagNullable = 2;
+        /// <summary>
+        /// For map types, whether the keys within each map value are sorted.
+        /// </summary>
+        public const long ArrowFlagMapKeysSorted = 4;
+
+        /// <summary>
+        /// Get the value of a particular flag.
+        /// </summary>
+        /// <remarks>
+        /// Known valid flags are <see cref="ArrowFlagDictionaryOrdered" />,
+        /// <see cref="ArrowFlagNullable" />, and <see cref="ArrowFlagMapKeysSorted" />.
+        /// </remarks>
+        public bool GetFlag(long flag)
+        {
+            return (flags & flag) == flag;
+        }
+
+        private static IntPtr ConstructChildren(IArrowType datatype, out long numChildren)
+        {
+            if (datatype is NestedType nestedType)
+            {
+                IReadOnlyList<Field> fields = nestedType.Fields;
+                int numFields = fields.Count;
+                numChildren = numFields;
+
+                unsafe
+                {
+                    IntPtr* pointerList = (IntPtr*)Marshal.AllocHGlobal(numFields * IntPtr.Size);
+
+                    for (var i = 0; i < numChildren; i++)
+                    {
+                        var cSchema = new CArrowSchema(fields[i]);
+                        IntPtr exportedSchema = cSchema.AllocateAsPtr();
+                        pointerList[i] = exportedSchema;
+                    }
+
+                    return (IntPtr)pointerList;
+                }
+
+            }
+            else
+            {
+                numChildren = 0;
+                return IntPtr.Zero;
+            }
+        }
+
+        private IntPtr GetChild(int i)
+        {
+            if (i >= n_children)
+            {
+                throw new Exception("Child index out of bounds.");
+            }
+            if (children == IntPtr.Zero)
+            {
+                throw new Exception("Children array is null.");
+            }
+            unsafe
+            {
+                return ((IntPtr*)children)[i];
+            }
+        }
+
+        private static IntPtr ConstructDictionary(IArrowType datatype)
+        {
+            if (datatype is DictionaryType dictType)
+            {
+                var cSchema = new CArrowSchema(dictType.ValueType);
+                return cSchema.AllocateAsPtr();
+            }
+            else
+            {
+                return IntPtr.Zero;
+            }
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as an Arrow type.
+        /// </summary>
+        /// <param name="datatype">The Arrow type to export.</param>
+        public CArrowSchema(IArrowType datatype)
```

Comment:
I'm a bit confused by the intention/use case here. I assumed the intention was so C# code could consume "C Data Interface" memory already in memory created by some other system. 

Is the use case here the other way? Maybe instead of putting these as methods on the `CArrowSchema` struct, it would be better to have a separate class/API/factory that created these. And then `CArrowSchema` struct would stay about defining the data structure, and not contain the logic of turning C# objects into the data structure.